README.md
# Blowfish ECB for Dart

<p align="center">
<a href="https://pub.dev/packages/blowfish_ecb"><img src="https://img.shields.io/pub/v/blowfish_ecb" alt="pub: blowfish_ecb"></a>
<a href="https://opensource.org/licenses/MIT"><img src="https://img.shields.io/badge/license-AGPL v3.0-green.svg" alt="License: AGPL"></a>
<a href="https://pub.dev/packages/lint"><img src="https://img.shields.io/badge/style-lint-4BC0F5.svg" alt="style: lint"></a>
</p>

A pure Dart [Codec](https://api.dart.dev/stable/2.10.4/dart-convert/Codec-class.html)
implementation for the [Blowfish ECB](https://www.schneier.com/academic/blowfish/)
encryption algorithm.

## Usage
The `BlowfishECB` class fully implements [Codec](https://api.dart.dev/stable/2.10.4/dart-convert/Codec-class.html).

The following simple usage is adapted from the included example project:
```dart
// Instantiate the codec with a key.
final blowfish = BlowfishECB(key);

// Encrypt the input data.
final encryptedData = blowfish.encode(message);

// Decrypt the encrypted data.
final decryptedData = blowfish.decode(encryptedData);
```

## License
Everything is licenced under the GNU Lesser General Public License v3 or above.  
See [`LICENCE`](LICENSE) and [`LICENCE.LESSER`](LICENSE.LESSER) for more
information.

Essentially, if this package is modified in your project, the modified package
sources must be released.

## Inspiration
The algorithm implementation was ported over from the
[Versile Python implementation](https://github.com/versiledev/versile-python/blob/master/versile/crypto/algorithm/blowfish.py).

CHANGELOG.md
## 0.1.0

- Stable null safety!

## 0.1.0-nullsafety.1

- Initial release


test/constants.dart
/*
Blowfish ECB dart:convert Codec implementation.
Copyright (C) 2021 hacker1024

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

/// "Hello, world!"
const sampleKey = [
  0x48,
  0x65,
  0x6c,
  0x6c,
  0x6f,
  0x2c,
  0x20,
  0x77,
  0x6f,
  0x72,
  0x6c,
  0x64,
  0x21,
];

/// Sample data for testing.
const sampleDecryptedData = [
  0x4c,
  0x6f,
  0x72,
  0x65,
  0x6d,
  0x20,
  0x69,
  0x70,
  0x73,
  0x75,
  0x6d,
  0x20,
  0x64,
  0x6f,
  0x6c,
  0x6f,
  0x72,
  0x20,
  0x73,
  0x69,
  0x74,
  0x20,
  0x61,
  0x6d,
  0x65,
  0x74,
  0x2c,
  0x20,
  0x63,
  0x6f,
  0x6e,
  0x73,
  0x65,
  0x63,
  0x74,
  0x65,
  0x74,
  0x75,
  0x72,
  0x20,
  0x61,
  0x64,
  0x69,
  0x70,
  0x69,
  0x73,
  0x63,
  0x69,
  0x6e,
  0x67,
  0x20,
  0x65,
  0x6c,
  0x69,
  0x74,
  0x2c,
  0x20,
  0x73,
  0x65,
  0x64,
  0x20,
  0x64,
  0x6f,
  0x20,
  0x65,
  0x69,
  0x75,
  0x73,
  0x6d,
  0x6f,
  0x64,
  0x20,
  0x74,
  0x65,
  0x6d,
  0x70,
  0x6f,
  0x72,
  0x20,
  0x69,
  0x6e,
  0x63,
  0x69,
  0x64,
  0x69,
  0x64,
  0x75,
  0x6e,
  0x74,
  0x20,
  0x75,
  0x74,
  0x20,
  0x6c,
  0x61,
  0x62,
  0x6f,
  0x72,
  0x65,
  0x20,
  0x65,
  0x74,
  0x20,
  0x64,
  0x6f,
  0x6c,
  0x6f,
  0x72,
  0x65,
  0x20,
  0x6d,
  0x61,
  0x67,
  0x6e,
  0x61,
  0x20,
  0x61,
  0x6c,
  0x69,
  0x71,
  0x75,
  0x61,
  0x2e,
  0x20,
  0x55,
  0x74,
  0x20,
  0x65,
  0x6e,
  0x69,
  0x6d,
  0x20,
  0x61,
  0x64,
  0x20,
  0x6d,
  0x69,
  0x6e,
  0x69,
  0x6d,
  0x20,
  0x76,
  0x65,
  0x6e,
  0x69,
  0x61,
  0x6d,
  0x2c,
  0x20,
  0x71,
  0x75,
  0x69,
  0x73,
  0x20,
  0x6e,
  0x6f,
  0x73,
  0x74,
  0x72,
  0x75,
  0x64,
  0x20,
  0x65,
  0x78,
  0x65,
  0x72,
  0x63,
  0x69,
  0x74,
  0x61,
  0x74,
  0x69,
  0x6f,
  0x6e,
  0x20,
  0x75,
  0x6c,
  0x6c,
  0x61,
  0x6d,
  0x63,
  0x6f,
  0x20,
  0x6c,
  0x61,
  0x62,
  0x6f,
  0x72,
  0x69,
  0x73,
  0x20,
  0x6e,
  0x69,
  0x73,
  0x69,
  0x20,
  0x75,
  0x74,
  0x20,
  0x61,
  0x6c,
  0x69,
  0x71,
  0x75,
  0x69,
  0x70,
  0x20,
  0x65,
  0x78,
  0x20,
  0x65,
  0x61,
  0x20,
  0x63,
  0x6f,
  0x6d,
  0x6d,
  0x6f,
  0x64,
  0x6f,
  0x20,
  0x63,
  0x6f,
  0x6e,
  0x73,
  0x65,
  0x71,
  0x75,
  0x61,
  0x74,
  0x2e,
  0x20,
  0x44,
  0x75,
  0x69,
  0x73,
  0x20,
  0x61,
  0x75,
  0x74,
  0x65,
  0x20,
  0x69,
  0x72,
  0x75,
  0x72,
  0x65,
  0x20,
  0x64,
  0x6f,
  0x6c,
  0x6f,
  0x72,
  0x20,
  0x69,
  0x6e,
  0x20,
  0x72,
  0x65,
  0x70,
  0x72,
  0x65,
  0x68,
  0x65,
  0x6e,
  0x64,
  0x65,
  0x72,
  0x69,
  0x74,
  0x20,
  0x69,
  0x6e,
  0x20,
  0x76,
  0x6f,
  0x6c,
  0x75,
  0x70,
  0x74,
  0x61,
  0x74,
  0x65,
  0x20,
  0x76,
  0x65,
  0x6c,
  0x69,
  0x74,
  0x20,
  0x65,
  0x73,
  0x73,
  0x65,
  0x20,
  0x63,
  0x69,
  0x6c,
  0x6c,
  0x75,
  0x6d,
  0x20,
  0x64,
  0x6f,
  0x6c,
  0x6f,
  0x72,
  0x65,
  0x20,
  0x65,
  0x75,
  0x20,
  0x66,
  0x75,
  0x67,
  0x69,
  0x61,
  0x74,
  0x20,
  0x6e,
  0x75,
  0x6c,
  0x6c,
  0x61,
  0x20,
  0x70,
  0x61,
  0x72,
  0x69,
  0x61,
  0x74,
  0x75,
  0x72,
  0x2e,
  0x20,
  0x45,
  0x78,
  0x63,
  0x65,
  0x70,
  0x74,
  0x65,
  0x75,
  0x72,
  0x20,
  0x73,
  0x69,
  0x6e,
  0x74,
  0x20,
  0x6f,
  0x63,
  0x63,
  0x61,
  0x65,
  0x63,
  0x61,
  0x74,
  0x20,
  0x63,
  0x75,
  0x70,
  0x69,
  0x64,
  0x61,
  0x74,
  0x61,
  0x74,
  0x20,
  0x6e,
  0x6f,
  0x6e,
  0x20,
  0x70,
  0x72,
  0x6f,
  0x69,
  0x64,
  0x65,
  0x6e,
  0x74,
  0x2c,
  0x20,
  0x73,
  0x75,
  0x6e,
  0x74,
  0x20,
  0x69,
  0x6e,
  0x20,
  0x63,
  0x75,
  0x6c,
  0x70,
  0x61,
  0x20,
  0x71,
  0x75,
  0x69,
  0x20,
  0x6f,
  0x66,
  0x66,
  0x69,
  0x63,
  0x69,
  0x61,
  0x20,
  0x64,
  0x65,
  0x73,
  0x65,
  0x72,
  0x75,
  0x6e,
  0x74,
  0x20,
  0x6d,
  0x6f,
  0x6c,
  0x6c,
  0x69,
  0x74,
  0x20,
  0x61,
  0x6e,
  0x69,
  0x6d,
  0x20,
  0x69,
  0x64,
  0x20,
  0x65,
  0x73,
  0x74,
  0x20,
  0x6c,
  0x61,
  0x62,
];

/// [sampleDecryptedData] encrypted with [sampleKey] using
/// http://sladex.org/blowfish.js/ (removed padding)
const sampleEncryptedData = [
  0xb5,
  0x95,
  0x34,
  0x03,
  0x4d,
  0x63,
  0xa9,
  0x0c,
  0x23,
  0x79,
  0x1d,
  0xe4,
  0x79,
  0x23,
  0x9d,
  0xc8,
  0x73,
  0xdb,
  0x03,
  0xca,
  0xfd,
  0x6d,
  0xc1,
  0x84,
  0xcd,
  0x96,
  0x30,
  0xe8,
  0xba,
  0x65,
  0x61,
  0xd4,
  0xaf,
  0xf9,
  0x4a,
  0xb9,
  0x44,
  0x97,
  0x5d,
  0x3f,
  0x25,
  0x93,
  0xe5,
  0x3d,
  0x33,
  0x41,
  0x6f,
  0x73,
  0xcf,
  0x88,
  0xfc,
  0xf1,
  0x0f,
  0x23,
  0x48,
  0xd9,
  0x84,
  0x76,
  0xa1,
  0x08,
  0x5f,
  0x1c,
  0x5c,
  0x09,
  0xa8,
  0x99,
  0x2d,
  0x6a,
  0xe5,
  0xb3,
  0x88,
  0x83,
  0x8a,
  0x79,
  0x64,
  0xc1,
  0x8a,
  0x53,
  0xec,
  0x88,
  0x70,
  0x34,
  0x91,
  0x9d,
  0xf1,
  0xd6,
  0x43,
  0x0d,
  0x50,
  0xec,
  0xe1,
  0x19,
  0xff,
  0x8f,
  0x41,
  0xb8,
  0x0a,
  0x89,
  0x3b,
  0x72,
  0xc9,
  0x35,
  0x9b,
  0x81,
  0x9b,
  0xbd,
  0x59,
  0x45,
  0x81,
  0x4d,
  0x33,
  0xd9,
  0x50,
  0xfc,
  0x23,
  0x2a,
  0xdb,
  0x6e,
  0xb1,
  0xcc,
  0x60,
  0x11,
  0xca,
  0x74,
  0x69,
  0x3e,
  0xe2,
  0x79,
  0xff,
  0xda,
  0xdb,
  0x1b,
  0x37,
  0xc8,
  0xe1,
  0x30,
  0x6e,
  0x45,
  0xdc,
  0x18,
  0xc0,
  0x16,
  0x9c,
  0x12,
  0x2f,
  0x9a,
  0x98,
  0x31,
  0x13,
  0x73,
  0xee,
  0xf9,
  0x68,
  0x3a,
  0x30,
  0x9e,
  0x5f,
  0x4e,
  0x56,
  0xd6,
  0xd3,
  0x64,
  0x94,
  0x01,
  0x69,
  0xea,
  0x06,
  0x35,
  0x9c,
  0x4e,
  0x18,
  0xc4,
  0x21,
  0xe2,
  0x0b,
  0xfd,
  0xc1,
  0x71,
  0xdd,
  0xc4,
  0x21,
  0xf6,
  0x5c,
  0x19,
  0x76,
  0x8c,
  0x54,
  0x2b,
  0x07,
  0xc2,
  0xb1,
  0xb2,
  0x1c,
  0xa4,
  0x65,
  0x2e,
  0x84,
  0x52,
  0xeb,
  0x04,
  0x40,
  0x9e,
  0x2e,
  0x17,
  0xc4,
  0xf4,
  0xc6,
  0xb5,
  0x3f,
  0xec,
  0xf2,
  0x72,
  0x3a,
  0x98,
  0x73,
  0x35,
  0xa6,
  0x07,
  0xf8,
  0x4b,
  0xda,
  0xdb,
  0xc9,
  0xa6,
  0x3d,
  0x10,
  0xce,
  0x44,
  0x39,
  0xcf,
  0xfd,
  0x7c,
  0xd8,
  0xfb,
  0x7c,
  0x19,
  0xcb,
  0xd5,
  0x8c,
  0xeb,
  0x47,
  0x6f,
  0x40,
  0x3d,
  0x69,
  0x20,
  0xb3,
  0xe5,
  0xfd,
  0xca,
  0xf2,
  0x80,
  0x59,
  0xf2,
  0x8b,
  0xbd,
  0xb8,
  0xb2,
  0xb8,
  0xbf,
  0xed,
  0x0c,
  0xa1,
  0x97,
  0x18,
  0x99,
  0x9c,
  0x40,
  0x8b,
  0xd8,
  0x30,
  0x8d,
  0xe3,
  0x81,
  0x99,
  0xf3,
  0x83,
  0xb1,
  0x95,
  0xa1,
  0x39,
  0x5b,
  0x20,
  0x72,
  0x27,
  0x90,
  0x9c,
  0x2c,
  0x72,
  0xfa,
  0xb5,
  0xac,
  0x31,
  0x0e,
  0x53,
  0xe6,
  0x41,
  0xde,
  0x0e,
  0x40,
  0x69,
  0x9e,
  0x64,
  0x97,
  0x33,
  0xe0,
  0x3d,
  0xec,
  0x63,
  0x42,
  0x12,
  0x68,
  0xcb,
  0x01,
  0xbb,
  0xaa,
  0x56,
  0xe9,
  0x17,
  0xa0,
  0xbf,
  0x02,
  0xd2,
  0x32,
  0x87,
  0xdc,
  0xc9,
  0x6c,
  0x35,
  0xd4,
  0xe7,
  0x8d,
  0x65,
  0x33,
  0x3a,
  0xe4,
  0x51,
  0x6d,
  0x56,
  0xb1,
  0x69,
  0x3a,
  0x01,
  0xb2,
  0x15,
  0xea,
  0x62,
  0x64,
  0xf4,
  0xd3,
  0xeb,
  0x1e,
  0xcc,
  0x03,
  0x2d,
  0x2e,
  0xc5,
  0x47,
  0xac,
  0xe5,
  0x86,
  0x0f,
  0xe5,
  0xa1,
  0x44,
  0x9a,
  0xaf,
  0x8f,
  0x10,
  0x60,
  0x6f,
  0xc5,
  0x3e,
  0x5a,
  0xee,
  0x65,
  0xe4,
  0x91,
  0x7f,
  0xc6,
  0x9e,
  0xd0,
  0xed,
  0xd0,
  0xc9,
  0xb2,
  0xf5,
  0x8a,
  0xdf,
  0x47,
  0x07,
  0x5f,
  0x04,
  0x60,
  0xa0,
  0xee,
  0xab,
  0x08,
  0x1e,
  0x82,
  0xa4,
  0x1d,
  0x5b,
  0xe1,
  0x14,
  0x73,
  0x6c,
  0x2d,
  0x9b,
  0x51,
  0x97,
  0xe2,
  0x09,
  0xee,
  0x94,
  0x02,
  0xc5,
  0x9d,
  0xea,
  0x05,
  0x17,
  0xa0,
  0x07,
  0x54,
  0xd4,
  0xc1,
  0xcc,
  0xd7,
  0xd5,
  0x22,
  0x4e,
  0x42,
  0x7b,
  0x45,
  0xd6,
  0xf4,
  0x80,
  0x49,
  0xbc,
  0xa0,
];


test/blowfish_ecb_test.dart
/*
Blowfish ECB dart:convert Codec implementation.
Copyright (C) 2021 hacker1024

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

import 'dart:async';
import 'dart:convert';
import 'dart:typed_data';

import 'package:blowfish_ecb/src/blowfish_ecb.dart';
import 'package:test/test.dart';

import 'constants.dart';

void main() {
  group(
    'Algorithm',
    () {
      late BlowfishECB blowfish;

      setUp(() => blowfish = BlowfishECB(Uint8List.fromList(sampleKey)));

      test(
        'Encryption',
        () {
          final encryptedData =
              blowfish.encode(Uint8List.fromList(sampleDecryptedData));

          expect(encryptedData, equals(sampleEncryptedData));
        },
      );

      test(
        'Decryption',
        () {
          final decryptedData =
              blowfish.decode(Uint8List.fromList(sampleEncryptedData));

          expect(decryptedData, equals(sampleDecryptedData));
        },
      );
    },
  );

  group(
    'Chunked conversions',
    () {
      Future<void> testChunkedConversion(
        Converter<List<int>, Uint8List> converter,
        List<int> input,
        List<int> expectedOutput,
      ) async {
        final completer = Completer<List<int>>();

        final inputSink = converter.startChunkedConversion(
          ChunkedConversionSink<Uint8List>.withCallback(
            (chunks) =>
                completer.complete([for (final chunk in chunks) ...chunk]),
          ),
        );
        // Add the data in two groups to simulate chunks going through.
        inputSink.add(input.sublist(0, 32));
        inputSink.add(input.sublist(32));
        inputSink.close();

        expect(await completer.future, equals(expectedOutput));
      }

      late BlowfishECB blowfish;

      setUp(() => blowfish = BlowfishECB(Uint8List.fromList(sampleKey)));

      test(
        'Encryption',
        () => testChunkedConversion(
          blowfish.encoder,
          sampleDecryptedData,
          sampleEncryptedData,
        ),
      );

      test(
        'Decryption',
        () => testChunkedConversion(
          blowfish.decoder,
          sampleEncryptedData,
          sampleDecryptedData,
        ),
      );
    },
  );

  group(
    'Validation',
    () {
      test(
        'Key size should be at most 56 bytes.',
        () {
          expect(
              () => BlowfishECB(Uint8List(56)), isNot(throwsFormatException));
          expect(() => BlowfishECB(Uint8List(57)), throwsFormatException);
        },
      );

      test(
        'Data must be padded to block sizes of 8',
        () {
          final blowfish = BlowfishECB(Uint8List.fromList(sampleKey));

          // Encoding
          expect(() => blowfish.encode(Uint8List(200)),
              isNot(throwsFormatException));
          expect(() => blowfish.encode(Uint8List(201)), throwsFormatException);

          // Decoding
          expect(() => blowfish.decode(Uint8List(200)),
              isNot(throwsFormatException));
          expect(() => blowfish.decode(Uint8List(201)), throwsFormatException);
        },
      );
    },
  );
}


lib/src/blowfish_ecb.dart
/*
Blowfish ECB dart:convert Codec implementation.
Copyright (C) 2021 hacker1024

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

// Loosely based on https://github.com/versiledev/versile-python/blob/master/versile/crypto/algorithm/blowfish.py

import 'dart:convert';
import 'dart:typed_data';

import 'package:blowfish_ecb/src/blowfish_ecb_converter.dart';
import 'package:blowfish_ecb/src/tables.dart';

/// A Blowfish ECB [Codec] implementation.
class BlowfishECB extends Codec<List<int>, List<int>> {
  /// The size of a block of Blowfish ECB data.
  ///
  /// All data must be a multiple of this number in length.
  static const blockSize = 8;

  final List<int> _p = copyP(pInit);
  final List<List<int>> _s = copyS(sInit);

  /// A [Converter] that encrypts data.
  @override
  Converter<List<int>, Uint8List> get encoder => BlowfishECBEncoder(_p, _s);

  /// A [Converter] that decrypts data.
  @override
  Converter<List<int>, Uint8List> get decoder => BlowfishECBDecoder(_p, _s);

  /// Creates an instance of the codec initialized with the given [key].
  BlowfishECB(Uint8List key) {
    final keyLength = key.length;

    if (keyLength > 56) {
      throw FormatException('Max key length is 448 bits (56 bytes)', key);
    }

    var j = 0;
    for (var i = 0; i < _p.length; ++i) {
      var data = 0;
      for (var k = 0; k < 4; ++k) {
        data = ((data << 8) & 0xffffffff) | key[j];
        ++j;
        if (j >= keyLength) j = 0;
      }
      _p[i] ^= data;
    }

    final data = Uint8List(8);
    for (var i = 0; i < _p.length; i += 2) {
      BlowfishECBEncoder.encryptBlock(data, 0, _p, _s);
      _p[i] = (data[0] << 24) + (data[1] << 16) + (data[2] << 8) + data[3];
      _p[i + 1] = (data[4] << 24) + (data[5] << 16) + (data[6] << 8) + data[7];
    }

    for (var i = 0; i < 4; ++i) {
      for (var j = 0; j < 256; j += 2) {
        BlowfishECBEncoder.encryptBlock(data, 0, _p, _s);
        _s[i][j] = (data[0] << 24) + (data[1] << 16) + (data[2] << 8) + data[3];
        _s[i][j + 1] =
            (data[4] << 24) + (data[5] << 16) + (data[6] << 8) + data[7];
      }
    }
  }
}


example/blowfish_ecb_example.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:blowfish_ecb/blowfish_ecb.dart';

void main() {
  const key = 'Passw0rd!';
  const message = 'Hello, world!';

  // Encode the key and instantiate the codec.
  final blowfish = BlowfishECB(Uint8List.fromList(utf8.encode(key)));

  // Encrypt the input (with padding to fit the 8-bit block size).
  print('Encrypting "$message" with PKCS #5 padding.');
  final encryptedData = blowfish.encode(padPKCS5(utf8.encode(message)));

  // Decrypt the encrypted data.
  print('Decrypting "${hexEncode(encryptedData)}".');
  var decryptedData = blowfish.decode(encryptedData);
  // Remove PKCS5 padding.
  decryptedData = decryptedData.sublist(
      0, decryptedData.length - getPKCS5PadCount(decryptedData));
  print('Got "${utf8.decode(decryptedData)}".');
}

String hexEncode(List<int> bytes) =>
    bytes.map((byte) => byte.toRadixString(16).padLeft(2, '0')).join();

Uint8List padPKCS5(List<int> input) {
  final inputLength = input.length;
  final paddingValue = 8 - (inputLength % 8);
  final outputLength = inputLength + paddingValue;

  final output = Uint8List(outputLength);
  for (var i = 0; i < inputLength; ++i) {
    output[i] = input[i];
  }
  output.fillRange(outputLength - paddingValue, outputLength, paddingValue);

  return output;
}

int getPKCS5PadCount(List<int> input) {
  if (input.length % 8 != 0) {
    throw FormatException('Block size is invalid!', input);
  }

  final count = input.last;
  final paddingStartIndex = input.length - count;
  for (var i = input.length - 1; i >= paddingStartIndex; --i) {
    if (input[i] != count) {
      throw const FormatException('Padding is not valid PKCS5 padding!');
    }
  }

  return count;
}


lib/src/blowfish_ecb_converter.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:blowfish_ecb/src/tables.dart';
import 'package:meta/meta.dart';

abstract class BlowfishECBConverter extends Converter<List<int>, Uint8List> {
  /// A base P list initialized with the key.
  final List<int> p;

  /// A base S list initialized with the key.
  final List<List<int>> s;

  const BlowfishECBConverter(this.p, this.s);

  /// Encrypts or decrypts the given [input], returning a new [Uint8List] with
  /// the output.
  @override
  Uint8List convert(List<int> input) {
    final result = Uint8List.fromList(input);
    _transform(
      result,
      copyP(p),
      copyS(s),
    );
    return result;
  }

  /// Starts a chunked conversion.
  ///
  /// The length of added chunks must be a multiple of 8.
  @override
  Sink<List<int>> startChunkedConversion(Sink<Uint8List> sink) {
    return _BlowfishECBConverterSink(
      sink,
      (block) => _transform(
        block,
        copyP(p),
        copyS(s),
      ),
    );
  }

  @protected
  void _transformBlock(
    Uint8List data,
    int startIndex,
    List<int> p,
    List<List<int>> s,
  );

  void _transform(
    Uint8List input,
    List<int> p,
    List<List<int>> s,
  ) {
    final length = input.length;
    if (length == 8) {
      _transformBlock(input, 0, p, s);
    } else {
      if (length % 8 != 0) {
        throw FormatException('Data not aligned with 8-byte blocksize', input);
      }

      for (var start = 0; start < length; start += 8) {
        _transformBlock(input, start, p, s);
      }
    }
  }

  static void _transformBlockCommon({
    required Uint8List data,
    required int startIndex,
    required List<int> p,
    required List<List<int>> s,
    required int loopStartAt,
    required int looopStopBefore,
    required int loopStep,
  }) {
    var bL = (data[0 + startIndex] << 24) +
        (data[1 + startIndex] << 16) +
        (data[2 + startIndex] << 8) +
        data[3 + startIndex];
    var bR = (data[4 + startIndex] << 24) +
        (data[5 + startIndex] << 16) +
        (data[6 + startIndex] << 8) +
        data[7 + startIndex];

    for (var i = loopStartAt; i != looopStopBefore; i += loopStep) {
      bL ^= p[i];
      bR ^= _feistel(bL, s);

      final swap = bL;
      bL = bR;
      bR = swap;
    }

    final swap = bL;
    bL = bR;
    bR = swap;

    bR ^= p[looopStopBefore];
    bL ^= p[looopStopBefore + loopStep];

    data[0 + startIndex] = bL >> 24;
    data[1 + startIndex] = bL >> 16;
    data[2 + startIndex] = bL >> 8;
    data[3 + startIndex] = bL;
    data[4 + startIndex] = bR >> 24;
    data[5 + startIndex] = bR >> 16;
    data[6 + startIndex] = bR >> 8;
    data[7 + startIndex] = bR;
  }

  static int _feistel(int x, List<List<int>> s) {
    var _x = x;
    final d = _x & 0xff;
    _x >>= 8;
    final c = _x & 0xff;
    _x >>= 8;
    final b = _x & 0xff;
    _x >>= 8;
    final a = _x & 0xff;
    var y = (s[0][a] + s[1][b]) & 0xffffffff;
    y ^= s[2][c];
    y = (y + s[3][d]) & 0xffffffff;
    return y;
  }
}

class _BlowfishECBConverterSink extends Sink<List<int>> {
  final Sink<Uint8List> output;
  final void Function(Uint8List block) transformBlock;

  _BlowfishECBConverterSink(this.output, this.transformBlock);

  @override
  void add(List<int> chunk) {
    final block = Uint8List.fromList(chunk);
    transformBlock(block);
    output.add(block);
  }

  @override
  void close() => output.close();
}

class BlowfishECBEncoder extends BlowfishECBConverter {
  const BlowfishECBEncoder(List<int> p, List<List<int>> s) : super(p, s);

  @override
  void _transformBlock(
    Uint8List data,
    int startIndex,
    List<int> p,
    List<List<int>> s,
  ) =>
      encryptBlock(data, startIndex, p, s);

  /// Encrypts an 8-byte block of data in an existing list of [data], starting
  /// at [startIndex].
  ///
  /// This is used internally by the package and should not need to be called in
  /// any other situation.
  static void encryptBlock(
    Uint8List data,
    int startIndex,
    List<int> p,
    List<List<int>> s,
  ) =>
      BlowfishECBConverter._transformBlockCommon(
        data: data,
        startIndex: startIndex,
        p: p,
        s: s,
        loopStartAt: 0,
        looopStopBefore: 16,
        loopStep: 1,
      );
}

class BlowfishECBDecoder extends BlowfishECBConverter {
  const BlowfishECBDecoder(List<int> p, List<List<int>> s) : super(p, s);

  @override
  void _transformBlock(
    Uint8List data,
    int startIndex,
    List<int> p,
    List<List<int>> s,
  ) =>
      _decryptBlock(data, startIndex, p, s);

  /// Decrypts an 8-byte block of data in an existing list of [data], starting
  /// at [startIndex].
  static void _decryptBlock(
    Uint8List data,
    int startIndex,
    List<int> p,
    List<List<int>> s,
  ) =>
      BlowfishECBConverter._transformBlockCommon(
        data: data,
        startIndex: startIndex,
        p: p,
        s: s,
        loopStartAt: 17,
        looopStopBefore: 1,
        loopStep: -1,
      );
}
